sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFnReflector.java

```diff
@@ -91,1012 +36,268 @@
+  private static final Map<Class<?>, DoFnSignature> SIGNATURE_CACHE = new LinkedHashMap<>();
 
-  /**
-   * @return the {@link DoFnReflector} for the given {@link DoFn}.
-   */
-  public static DoFnReflector of(
+  /** @return the {@link DoFnReflector} for the given {@link DoFn}. */
+  static synchronized DoFnSignature getSignature(
       @SuppressWarnings("rawtypes") Class<? extends DoFn> fn) {
-    DoFnReflector reflector = REFLECTOR_CACHE.get(fn);
-    if (reflector != null) {
-      return reflector;
+    DoFnSignature signature = SIGNATURE_CACHE.get(fn);
+    if (signature != null) {
+      return signature;
     }
 
-    reflector = new GenericDoFnReflector(fn);
-    REFLECTOR_CACHE.put(fn, reflector);
-    return reflector;
+    signature = getSignatureImpl(fn);
+    SIGNATURE_CACHE.put(fn, signature);
+    return signature;
   }
 
-  /**
-   * Create a {@link OldDoFn} that the {@link DoFn}.
-   */
-  public <InputT, OutputT> OldDoFn<InputT, OutputT> toDoFn(DoFn<InputT, OutputT> fn) {
-    if (usesSingleWindow()) {
-      return new WindowDoFnAdapter<InputT, OutputT>(this, fn);
-    } else {
-      return new SimpleDoFnAdapter<InputT, OutputT>(this, fn);
-    }
-  }
+  /** Analyzes a given {@link DoFn} class and extracts its {@link DoFnSignature}. */
+  private static DoFnSignature getSignatureImpl(Class<? extends DoFn> fnClass) {
+    TypeToken<?> inputT = null;
+    TypeToken<?> outputT = null;
 
-  private static String formatType(TypeToken<?> t) {
-    return ReflectHelpers.TYPE_SIMPLE_DESCRIPTION.apply(t.getType());
+    // Extract the input and output type.
+    if (!DoFn.class.isAssignableFrom(fnClass)) {
+      throw new IllegalArgumentException(
+          String.format("%s must be subtype of DoFn", fnClass.getSimpleName()));
+    }
+    for (TypeToken<?> supertype : TypeToken.of(fnClass).getTypes()) {
+      if (!supertype.getRawType().equals(DoFn.class)) {
+        continue;
+      }
+      Type[] args = ((ParameterizedType) supertype.getType()).getActualTypeArguments();
+      inputT = TypeToken.of(args[0]);
+      outputT = TypeToken.of(args[1]);
+    }
+    assert inputT != null;
+
+    Method processElementMethod = findAnnotatedMethod(DoFn.ProcessElement.class, fnClass, true);
+    Method startBundleMethod = findAnnotatedMethod(DoFn.StartBundle.class, fnClass, false);
+    Method finishBundleMethod = findAnnotatedMethod(DoFn.FinishBundle.class, fnClass, false);
+
+    return new DoFnSignature(
+        inputT,
+        outputT,
+        analyzeProcessElementMethod(processElementMethod, inputT, outputT),
+        (startBundleMethod == null)
+            ? null
+            : analyzeBundleMethod(startBundleMethod, inputT, outputT),
+        (finishBundleMethod == null)
+            ? null
+            : analyzeBundleMethod(finishBundleMethod, inputT, outputT));
   }
 
-  private static String format(Method m) {
-    return ReflectHelpers.CLASS_AND_METHOD_FORMATTER.apply(m);
+  private static <InputT, OutputT>
```

Comment:
`/** Construct the type token for the process context given the input and output type token. */`

May also be worth mentoining that it is in a helper method so that we can plumb generics appropriately.